akka-distributed-data/src/main/scala/akka/cluster/ddata/Gauge.scala

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.ddata
+
+import akka.cluster.ClusterEvent.CurrentClusterState
+import akka.cluster.{ MemberStatus, UniqueAddress }
+
+import scala.collection.JavaConverters._
+
+object Gauge {
+  val empty: Gauge = new Gauge
+  def apply(): Gauge = empty
+
+  /**
+   * Java API
+   */
+  def create(): Gauge = empty
+
+  private val Zero = BigInt(0)
+}
+
+/**
+ * Implements a gauge CRDT.
+ *
+ * A gauge CRDT has a value for each node in the cluster. These values can then be aggregated, with a sum, average,
+ * minimum or maximum.
+ *
+ * The gauge aggregations are calculated on request by supplying the current cluster state, along with the set of
+ * member statuses that are allowed to take part in the aggregation.
+ */
+@SerialVersionUID(1L)
+final case class Gauge private[akka] (
+    private[akka] val state: Map[UniqueAddress, (BigInt, BigInt)] = Map.empty,
+    override val delta: Option[Gauge] = None)
+    extends NodeVector[(BigInt, BigInt)](state)
+    with ReplicatedDataSerialization {
```

Comment:
How is the Gauge different from PNCounter, aside from that it's implemented differently?
Could we add these features to PNCounter?

Reply:
They could be the same, however it would fundamentally change how `PNCounter` is implemented, since `PNCounter` is not a `NodeVector`, it's two `GCounter`'s. The operations on `Gauge` (`min`, `max`, `average` etc) require knowing the PN summed value of each node individually, which excludes implementing it using two `GCounter`'s. I'm not sure if there's any benefits to one or the other implementations.